File: sdk/core/android/src/main/kotlin/telemetry/TelemetryCollector.kt
package com.rivalapexmediation.sdk.telemetry

import android.content.Context
import com.rivalapexmediation.sdk.BuildConfig
import com.rivalapexmediation.sdk.SDKConfig
import com.rivalapexmediation.sdk.models.EventType
import com.rivalapexmediation.sdk.models.TelemetryEvent
import com.google.gson.Gson
import com.rivalapexmediation.sdk.util.ClockProvider
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.ByteArrayOutputStream
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.zip.GZIPOutputStream

/**
 * Telemetry collector with batching and compression
 * 
 * Features:
 * - Event batching (reduces network calls)
 * - GZIP compression
 * - Retry logic with exponential backoff
 * - Background processing (no ANR risk)
 */
class TelemetryCollector(
    private val context: Context,
    private val config: SDKConfig
) {
    private val eventQueue = mutableListOf<TelemetryEvent>()
    private val queueLock = Any()
    
    private val executor = Executors.newSingleThreadScheduledExecutor { r ->
        Thread(r, "RivalApexMediation-Telemetry").apply {
            priority = Thread.MIN_PRIORITY
        }
    }
    
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()
    
    private val gson = Gson()
    
    private val batchSize = 10
    private val flushInterval = 30000L // 30 seconds
    private val random = java.util.Random()
    
    @Volatile
    private var isRunning = false
    private val sampleRate: Double get() = config.observabilitySampleRate.coerceIn(0.0, 1.0)

    // --- P1.9: Lightweight in-memory observability metrics (privacy-clean) ---
    // We track per {placement, adapter} rolling latencies and outcome counters in-memory
    // for quick developer diagnostics. Bounded memory via small reservoirs per key.
    private data class Key(val placement: String, val adapter: String)
    private data class Counters(
        var fills: Long = 0,
        var noFills: Long = 0,
        var timeouts: Long = 0,
        var errors: Long = 0,
    )
    private val counters = java.util.concurrent.ConcurrentHashMap<Key, Counters>()
    private val latencyReservoirs = java.util.concurrent.ConcurrentHashMap<Key, ArrayDeque<Long>>()
    private val reservoirMax = 200 // up to 200 recent latencies per key

    private fun getOrCreateCounters(key: Key): Counters {
        var c = counters[key]
        if (c != null) return c
        val created = Counters()
        val prev = counters.putIfAbsent(key, created)
        return prev ?: created
    }

    private fun getOrCreateReservoir(key: Key): ArrayDeque<Long> {
        var dq = latencyReservoirs[key]
        if (dq != null) return dq
        val created = ArrayDeque<Long>()
        val prev = latencyReservoirs.putIfAbsent(key, created)
        return prev ?: created
    }

    private fun recordOutcomeSampled(placement: String, adapter: String, outcome: String, latencyMs: Long?) {
        if (!config.observabilityEnabled) return
        if (!shouldSample()) return
        val key = Key(placement, adapter)
        // Update counters
        val c = getOrCreateCounters(key)
        when (outcome) {
            "fill" -> c.fills++
            "no_fill" -> c.noFills++
            "timeout" -> c.timeouts++
            "error" -> c.errors++
        }
        // Update latency reservoir
        if (latencyMs != null && latencyMs >= 0) {
            val dq = getOrCreateReservoir(key)
            dq.addLast(latencyMs)
            while (dq.size > reservoirMax) dq.removeFirst()
        }
    }

    // Percentiles computed over the local reservoir (best-effort, dev-only diagnostics)
    fun getLocalPercentiles(placement: String, adapter: String): Map<String, Long> {
        val key = Key(placement, adapter)
        val arr = latencyReservoirs[key]?.toLongArray() ?: LongArray(0)
        if (arr.isEmpty()) return emptyMap()
        java.util.Arrays.sort(arr)
        fun pct(p: Double): Long {
            if (arr.isEmpty()) return 0L
            val idx = kotlin.math.min(arr.size - 1, kotlin.math.max(0, kotlin.math.floor(p * (arr.size - 1)).toInt()))
            return arr[idx]
        }
        return mapOf(
            "p50" to pct(0.50),
            "p95" to pct(0.95),
            "p99" to pct(0.99),
        )
    }

    fun getLocalCounters(placement: String, adapter: String): Map<String, Long> {
        val key = Key(placement, adapter)
        val c = counters[key] ?: return emptyMap()
        return mapOf(
            "fills" to c.fills,
            "no_fills" to c.noFills,
            "timeouts" to c.timeouts,
            "errors" to c.errors,
        )
    }
    
    /**
     * Start telemetry collection
     */
    fun start() {
        if (!config.telemetryEnabled) {
            return
        }
        
        isRunning = true
        
        // Schedule periodic flush
        executor.scheduleWithFixedDelay(
            { flushEvents() },
            flushInterval,
            flushInterval,
            TimeUnit.MILLISECONDS
        )
    }
    
    /**
     * Stop telemetry collection
     */
    fun stop() {
        isRunning = false
        
        // Flush remaining events
        flushEvents()
        
        executor.shutdown()
        try {
            executor.awaitTermination(5, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
            executor.shutdownNow()
        }
    }
    
    /**
     * Record SDK initialization
     */
    fun recordInitialization() {
        recordEvent(TelemetryEvent(eventType = EventType.SDK_INIT))
    }
    
    /**
     * Record ad load event
     */
    fun recordAdLoad(placement: String, latency: Long, success: Boolean) {
        val eventType = if (success) EventType.AD_LOADED else EventType.AD_FAILED
        recordEvent(
            TelemetryEvent(
                eventType = eventType,
                placement = placement,
                latency = latency
            )
        )
    }
    
    /**
     * Record timeout event
     */
    fun recordTimeout(placement: String, reason: String) {
        recordEvent(
            TelemetryEvent(
                eventType = EventType.TIMEOUT,
                placement = placement,
                metadata = mapOf("reason" to reason)
            )
        )
    }
    
    /**
     * Record error event
     */
    fun recordError(errorCode: String, error: Throwable) {
        recordEvent(
            TelemetryEvent(
                eventType = EventType.AD_FAILED,
                errorCode = errorCode,
                errorMessage = error.message,
                metadata = mapOf(
                    "stack_trace" to error.stackTraceToString().take(500)
                )
            )
        )
    }
    
    /**
     * Record ANR detection
     */
    fun recordANR(threadName: String, stackTrace: String) {
        recordEvent(
            TelemetryEvent(
                eventType = EventType.ANR_DETECTED,
                metadata = mapOf(
                    "thread" to threadName,
                    "stack_trace" to stackTrace.take(500)
                )
            )
        )
    }

    /** Lightweight status counter for OM SDK availability/config state. */
    fun recordOmSdkStatus(status: String) {
        recordEvent(
            TelemetryEvent(
                eventType = EventType.OMSDK_STATUS,
                metadata = mapOf("status" to status)
            )
        )
    }

    /**
     * Adapter span start (observability). Uses sampling; no secrets allowed.
     */
    fun recordAdapterSpanStart(traceId: String, placement: String, adapter: String, metadata: Map<String, Any> = emptyMap()) {
        if (!config.observabilityEnabled) return
        if (!shouldSample()) return
        val safeMeta = HashMap<String, Any>()
        safeMeta["trace_id"] = traceId
        safeMeta["phase"] = "start"
        for ((k, v) in metadata) {
            if (k.length <= 40) safeMeta[k] = v
        }
        recordEvent(
            TelemetryEvent(
                eventType = EventType.ADAPTER_SPAN_START,
                placement = placement,
                networkName = adapter,
                metadata = safeMeta
            )
        )
    }

    /**
     * Adapter span finish (observability). Outcome taxonomy: fill | no_fill | timeout | error.
     * Includes latency and normalized error code/message when applicable. Uses sampling.
     */
    fun recordAdapterSpanFinish(
        traceId: String,
        placement: String,
        adapter: String,
        outcome: String,
        latencyMs: Long?,
        errorCode: String? = null,
        errorMessage: String? = null,
        metadata: Map<String, Any> = emptyMap()
    ) {
        if (!config.observabilityEnabled) return
        // Update local metrics using the same sampling gate
        recordOutcomeSampled(placement, adapter, outcome, latencyMs)
        if (!shouldSample()) return
        val safeMeta = HashMap<String, Any>()
        safeMeta["trace_id"] = traceId
        safeMeta["phase"] = "finish"
        safeMeta["outcome"] = outcome
        if (latencyMs != null) safeMeta["latency_ms"] = latencyMs
        // Merge limited caller-provided metadata (must be non-sensitive)
        for ((k, v) in metadata) {
            if (k.length <= 40) safeMeta[k] = v
        }
        recordEvent(
            TelemetryEvent(
                eventType = EventType.ADAPTER_SPAN_FINISH,
                placement = placement,
                networkName = adapter,
                latency = latencyMs,
                errorCode = errorCode,
                errorMessage = errorMessage?.take(200),
                metadata = safeMeta
            )
        )
    }

    /**
     * Record a credential validation success for a given network (BYO ValidationMode).
     * Metadata must not contain secrets; only include key names or booleans.
     */
    fun recordCredentialValidationSuccess(network: String, metadata: Map<String, Any> = emptyMap()) {
        recordEvent(
            TelemetryEvent(
                eventType = EventType.CREDENTIAL_VALIDATION_SUCCESS,
                networkName = network,
                metadata = metadata
            )
        )
    }

    /**
     * Record a credential validation failure for a given network (BYO ValidationMode).
     * Provide a normalized error code and optional message. No secrets allowed in metadata.
     */
    fun recordCredentialValidationFailure(network: String, code: String, message: String?, metadata: Map<String, Any> = emptyMap()) {
        recordEvent(
            TelemetryEvent(
                eventType = EventType.CREDENTIAL_VALIDATION_FAILED,
                networkName = network,
                errorCode = code,
                errorMessage = message,
                metadata = metadata
            )
        )
    }
    
    /**
     * Add event to queue
     */
    private fun recordEvent(event: TelemetryEvent) {
        if (!config.telemetryEnabled || !isRunning) {
            return
        }
        
        val safeEvent = TelemetryRedactor.sanitize(event)
        synchronized(queueLock) {
            eventQueue.add(safeEvent)
            
            // Auto-flush if batch size reached
            if (eventQueue.size >= batchSize) {
                executor.execute { flushEvents() }
            }

            // Apply hard cap to queue size to avoid memory growth
            val maxQ = (if (config.observabilityMaxQueue > 0) config.observabilityMaxQueue else 500).coerceAtLeast(100)
            if (eventQueue.size > maxQ) {
                // Drop oldest beyond cap
                eventQueue.subList(0, eventQueue.size - maxQ).clear()
            }
        }
    }
    
    /**
     * Flush events to backend
     */
    private fun flushEvents() {
        val eventsToSend: List<TelemetryEvent>
        
        synchronized(queueLock) {
            if (eventQueue.isEmpty()) {
                return
            }
            
            eventsToSend = eventQueue.toList()
            eventQueue.clear()
        }
        
        try {
            sendEvents(eventsToSend)
        } catch (e: Exception) {
            // Re-queue events on failure
            synchronized(queueLock) {
                eventQueue.addAll(0, eventsToSend)
                // Limit queue size to prevent memory issues (configurable)
                val maxQ = (if (config.observabilityMaxQueue > 0) config.observabilityMaxQueue else 500).coerceAtLeast(100)
                if (eventQueue.size > maxQ) {
                    eventQueue.subList(0, eventQueue.size - maxQ).clear()
                }
            }
        }
    }
    
    /**
     * Send events to telemetry endpoint
     */
    private fun sendEvents(events: List<TelemetryEvent>) {
        val payload = mapOf(
            "app_id" to config.appId,
            "sdk_version" to BuildConfig.SDK_VERSION,
            "platform" to "android",
            "events" to events
        )
        
        val json = gson.toJson(payload)
        val compressed = compressGzip(json.toByteArray())
        
        val request = Request.Builder()
            .url("${config.configEndpoint}/v1/telemetry")
            .post(compressed.toRequestBody("application/json".toMediaType()))
            .addHeader("Content-Encoding", "gzip")
            .addHeader("User-Agent", "RivalApexMediation-Android/${BuildConfig.SDK_VERSION}")
            .build()
        
        val response = httpClient.newCall(request).execute()
        
        if (!response.isSuccessful) {
            throw Exception("Telemetry upload failed: ${response.code}")
        }
    }
    
    /**
     * Compress data with GZIP
     */
    private fun compressGzip(data: ByteArray): ByteArray {
        val outputStream = ByteArrayOutputStream()
        GZIPOutputStream(outputStream).use { gzip ->
            gzip.write(data)
        }
        return outputStream.toByteArray()
    }

    private fun shouldSample(): Boolean {
        val p = sampleRate
        if (p >= 1.0) return true
        if (p <= 0.0) return false
        return random.nextDouble() < p
    }
}

/**
 * Performance metrics tracker
 */
class PerformanceMetrics {
    data class Metric(
        val name: String,
        val value: Double,
        val timestamp: Long = ClockProvider.clock.now()
    )
    
    private val metrics = mutableListOf<Metric>()
    
    fun record(name: String, value: Double) {
        synchronized(metrics) {
            metrics.add(Metric(name, value))
            
            // Keep only last 100 metrics
            if (metrics.size > 100) {
                metrics.removeAt(0)
            }
        }
    }
    
    fun getMetrics(): List<Metric> {
        return synchronized(metrics) {
            metrics.toList()
        }
    }
    
    fun clear() {
        synchronized(metrics) {
            metrics.clear()
        }
    }
}

File: sdk/core/ios/Sources/Telemetry/TelemetryCollector.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif
#if canImport(Compression)
import Compression
#endif

/// Telemetry collector with batching and compression
///
/// Features:
/// - Event batching (reduces network calls)
/// - GZIP compression
/// - Retry logic with exponential backoff
/// - Background processing (no main thread blocking)
public class TelemetryCollector: @unchecked Sendable {
    private let config: SDKConfig
    private let urlSession: URLSession
    
    private var eventQueue: [TelemetryEvent] = []
    private let queueLock = NSLock()
    
    private let backgroundQueue = DispatchQueue(label: "com.rivalapexmediation.telemetry", qos: .utility)
    private var flushTimer: DispatchSourceTimer?
    
    private let batchSize = 10
    private let flushInterval: TimeInterval = 30.0 // 30 seconds
    
    private var isRunning = false
    
    /// Initialize telemetry collector
    public init(config: SDKConfig) {
        self.config = config
        
        let configuration = URLSessionConfiguration.apexDefault()
        configuration.timeoutIntervalForRequest = 10
        configuration.timeoutIntervalForResource = 30
        self.urlSession = URLSession(configuration: configuration)
    }
    
    /// Start telemetry collection
    public func start() {
        guard config.telemetryEnabled else { return }
        
        isRunning = true
        
        // Schedule periodic flush
        let timer = DispatchSource.makeTimerSource(queue: backgroundQueue)
        timer.schedule(deadline: .now() + flushInterval, repeating: flushInterval)
        timer.setEventHandler { [weak self] in
            self?.flushEvents()
        }
        timer.resume()
        
        flushTimer = timer
    }
    
    /// Stop telemetry collection
    public func stop() {
        isRunning = false
        
        // Cancel timer
        flushTimer?.cancel()
        flushTimer = nil
        
        // Flush remaining events
        flushEvents()
    }
    
    // MARK: - Event Recording
    
    /// Record SDK initialization
    public func recordInitialization() {
        recordEvent(TelemetryEvent(eventType: .sdkInit))
    }
    
    /// Record ad load event
    public func recordAdLoad(placement: String, adType: AdType, networkName: String, latency: Int, success: Bool) {
        let eventType: EventType = success ? .adLoaded : .adFailed
        recordEvent(
            TelemetryEvent(
                eventType: eventType,
                placement: placement,
                adType: adType,
                networkName: networkName,
                latency: latency
            )
        )
    }
    
    /// Record ad impression
    public func recordImpression(placement: String, adType: AdType, networkName: String) {
        recordEvent(
            TelemetryEvent(
                eventType: .adImpression,
                placement: placement,
                adType: adType,
                networkName: networkName
            )
        )
    }
    
    /// Record ad click
    public func recordClick(placement: String, adType: AdType, networkName: String) {
        recordEvent(
            TelemetryEvent(
                eventType: .adClicked,
                placement: placement,
                adType: adType,
                networkName: networkName
            )
        )
    }
    
    /// Record timeout event
    public func recordTimeout(placement: String, adType: AdType, reason: String) {
        recordEvent(
            TelemetryEvent(
                eventType: .timeout,
                placement: placement,
                adType: adType,
                metadata: ["reason": reason]
            )
        )
    }
    
    /// Record error event
    public func recordError(errorCode: String, error: Error) {
        recordEvent(
            TelemetryEvent(
                eventType: .adFailed,
                errorCode: errorCode,
                errorMessage: error.localizedDescription,
                metadata: [
                    "error_domain": (error as NSError).domain,
                    "error_code": String((error as NSError).code)
                ]
            )
        )
    }
    
    /// Record ANR detection (not applicable on iOS, but kept for parity)
    public func recordANR(threadName: String, stackTrace: String) {
        recordEvent(
            TelemetryEvent(
                eventType: .anrDetected,
                metadata: [
                    "thread": threadName,
                    "stack_trace": String(stackTrace.prefix(500))
                ]
            )
        )
    }
    
    // MARK: - Private Methods
    
    /// Add event to queue
    private func recordEvent(_ event: TelemetryEvent) {
        guard config.telemetryEnabled, isRunning else { return }
        
        queueLock.lock()
        eventQueue.append(event)
        let shouldFlush = eventQueue.count >= batchSize
        queueLock.unlock()
        
        // Auto-flush if batch size reached
        if shouldFlush {
            backgroundQueue.async { [weak self] in
                self?.flushEvents()
            }
        }
    }
    
    /// Flush events to backend
    private func flushEvents() {
        queueLock.lock()
        guard !eventQueue.isEmpty else {
            queueLock.unlock()
            return
        }
        
        let eventsToSend = eventQueue
        eventQueue.removeAll()
        queueLock.unlock()
        
        Task.detached(priority: .utility) { [weak self] in
            guard let self else { return }
            do {
                try await self.sendEvents(eventsToSend)
            } catch {
                self.backgroundQueue.async { [weak self] in
                    guard let self else { return }
                    self.queueLock.lock()
                    self.eventQueue.insert(contentsOf: eventsToSend, at: 0)
                    
                    // Limit queue size to prevent memory issues
                    if self.eventQueue.count > 1000 {
                        self.eventQueue.removeFirst(self.eventQueue.count - 1000)
                    }
                    self.queueLock.unlock()
                }
            }
        }
    }
    
    /// Send events to telemetry endpoint
    private func sendEvents(_ events: [TelemetryEvent]) async throws {
        let payload: [String: Any] = [
            "app_id": config.appId,
            "sdk_version": "1.0.0",
            "platform": "ios",
            "events": events.map { event in
                var dict: [String: Any] = [
                    "event_type": event.eventType.rawValue,
                    "timestamp": ISO8601DateFormatter().string(from: event.timestamp)
                ]
                
                if let placement = event.placement { dict["placement"] = placement }
                if let adType = event.adType { dict["ad_type"] = adType.rawValue }
                if let networkName = event.networkName { dict["network_name"] = networkName }
                if let latency = event.latency { dict["latency"] = latency }
                if let errorCode = event.errorCode { dict["error_code"] = errorCode }
                if let errorMessage = event.errorMessage { dict["error_message"] = errorMessage }
                if let metadata = event.metadata { dict["metadata"] = metadata }
                
                return dict
            }
        ]
        
        let jsonData = try JSONSerialization.data(withJSONObject: payload)
        let compressed = try compress(jsonData)
        
        let url = URL(string: "\(config.configEndpoint)/v1/telemetry")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = compressed
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        #if canImport(Compression)
        request.addValue("gzip", forHTTPHeaderField: "Content-Encoding")
        #endif
        request.addValue("RivalApexMediation-iOS/1.0.0", forHTTPHeaderField: "User-Agent")
        
        let (_, response) = try await urlSession.apexData(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw TelemetryError.uploadFailed
        }
    }
    
    /// Compress data with GZIP
    private func compress(_ data: Data) throws -> Data {
        #if canImport(Compression)
        var compressedData = Data()
        
        try data.withUnsafeBytes { (sourceBuffer: UnsafeRawBufferPointer) in
            guard let sourcePtr = sourceBuffer.baseAddress else {
                throw TelemetryError.compressionFailed
            }
            
            let destinationBufferSize = data.count
            let destinationBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: destinationBufferSize)
            defer { destinationBuffer.deallocate() }
            
            let compressedSize = compression_encode_buffer(
                destinationBuffer,
                destinationBufferSize,
                sourcePtr.assumingMemoryBound(to: UInt8.self),
                data.count,
                nil,
                COMPRESSION_ZLIB
            )
            
            if compressedSize == 0 {
                throw TelemetryError.compressionFailed
            }
            
            compressedData = Data(bytes: destinationBuffer, count: compressedSize)
        }
        
        return compressedData
        #else
        return data
        #endif
    }
}

// MARK: - Performance Metrics

/// Performance metrics tracker
public class PerformanceMetrics {
    public struct Metric {
        public let name: String
        public let value: Double
        public let timestamp: Date
        
        public init(name: String, value: Double, timestamp: Date = Date()) {
            self.name = name
            self.value = value
            self.timestamp = timestamp
        }
    }
    
    private var metrics: [Metric] = []
    private let lock = NSLock()
    
    /// Record a metric
    public func record(name: String, value: Double) {
        lock.lock()
        metrics.append(Metric(name: name, value: value))
        
        // Keep only last 100 metrics
        if metrics.count > 100 {
            metrics.removeFirst()
        }
        lock.unlock()
    }
    
    /// Get all metrics
    public func getMetrics() -> [Metric] {
        lock.lock()
        defer { lock.unlock() }
        return metrics
    }
    
    /// Clear all metrics
    public func clear() {
        lock.lock()
        metrics.removeAll()
        lock.unlock()
    }
}

// MARK: - Errors

/// Telemetry errors
public enum TelemetryError: Error, LocalizedError {
    case uploadFailed
    case compressionFailed
    
    public var errorDescription: String? {
        switch self {
        case .uploadFailed:
            return "Failed to upload telemetry data"
        case .compressionFailed:
            return "Failed to compress telemetry data"
        }
    }
}
